### ISSUE TITLE
POST events to non-master node with chunked body forwarded to master without Content-Type

### ISSUE THREAD

**Issue by neoeinstein** [EventStore/EventStore#1574]
We have noticed that, when a slave node receives a post events request over the HTTP API with a chunked body, it is unable to properly forward the request to the master node. Chunked bodies are accepted when posted directly to the master, but not when forwarded. The reason seems to be due to the check in [this area of the HttpSendService](https://github.com/EventStore/EventStore/blob/daddeca886f43e40489efd7d9f45cb6ddd4aed8a/src/EventStore.Core/Services/HttpSendService.cs#L221) which fails to pass on the `Content-Type` header if a `Content-Length` header is not included, which it never is when the `Transfer-Encoding` is `chunked`.

Logs below from both EventStore and the Varnish proxy to demonstrate the issue.

Workaround is to always provide a fixed-sized body with a `Content-Length` header and no `Transfer-Encoding` header.

Logs from EventStore with `LogHttpRequests` set to `true`:
```
[00001,07,18:51:10.841] HTTP Request Received
03/09/2018 18:51:10
From: {varnish}:43428
POST http://{master}:2113/streams/test-1
Accept: application/vnd.eventstore.atom+json; charset=utf-8
X-Cloud-Trace-Context: ZZZZZZZ2/12345566
Via: 1.1 google
X-Forwarded-Proto: https
X-Forwarded-For: 8.X.X.X, 8.X.X.X, {lb}
ES-TrustedAuth: anonymous
X-Varnish: 131334
X-Forwarded-Host: host.es.com:2114
Content-Length: 0
Host: {master}
 
[00001,07,18:51:10.841] HTTP Response
03/09/2018 18:51:10
415 Invalid or missing Content-Type
Accept: application/vnd.eventstore.atom+json; charset=utf-8
X-Cloud-Trace-Context: ZZZZZZZ2/12345566
Via: 1.1 google
X-Forwarded-Proto: https
X-Forwarded-For: 8.X.X.X, 8.X.X.X, {lb}
ES-TrustedAuth: anonymous
X-Varnish: 131334
X-Forwarded-Host: host.es.com:2114
Content-Length: 0
Host: {master}
```

Logs below, IPs have been anonymized.
varnish log from a request passed directly to the master node:
```
     98775 Begin          c sess 0 HTTP/1
     98775 SessOpen       c {lb} 61133 a0 {varnish} 80 1520621471.085619 22
     98775 Link           c req 98776 rxreq
     98777 Begin          b bereq 98776 pass
     98777 Timestamp      b Start: 1520621471.085813 0.000000 0.000000
     98777 BereqMethod    b POST
     98777 BereqURL       b /streams/test-1
     98777 BereqProtocol  b HTTP/1.1
     98777 BereqHeader    b Host: host.es.com
     98777 BereqHeader    b Content-Type: application/vnd.eventstore.events+json; charset=utf-8
     98777 BereqHeader    b Accept: application/vnd.eventstore.atom+json; charset=utf-8
     98777 BereqHeader    b X-Cloud-Trace-Context: ZZZZZZZ1/12342541254
     98777 BereqHeader    b Via: 1.1 google
     98777 BereqHeader    b X-Forwarded-Proto: https
     98777 BereqHeader    b X-Forwarded-For: 8.X.X.X, 8.X.X.X, {lb}
     98777 BereqHeader    b ES-TrustedAuth: anonymous
     98777 BereqHeader    b X-Varnish: 98777
     98777 VCL_call       b BACKEND_FETCH
     98777 VCL_return     b fetch
     98777 BackendOpen    b 29 boot.es_dir({ES_MASTER) {ES_MASTER} 2114 {varnish} 55396
     98777 BereqHeader    b Transfer-Encoding: chunked
     98777 BackendStart   b {ES_MASTER} 2114
     98777 Timestamp      b Bereq: 1520621471.087000 0.001187 0.001187
     98777 Timestamp      b Beresp: 1520621471.122937 0.037124 0.035937
     98777 BerespProtocol b HTTP/1.1
     98777 BerespStatus   b 201
     98777 BerespReason   b Created
     98777 BerespHeader   b Access-Control-Allow-Methods: POST, DELETE, GET, OPTIONS
     98777 BerespHeader   b Access-Control-Allow-Headers: Content-Type, X-Requested-With, X-Forwarded-Host, X-Forwarded-Prefix, X-PINGOTHER, Authorization, ES-LongPoll, ES-ExpectedVersion, ES-EventId, ES-EventType, ES-RequiresMaster, ES-HardDelete, ES-ResolveLinkTos
     98777 BerespHeader   b Access-Control-Allow-Origin: *
     98777 BerespHeader   b Access-Control-Expose-Headers: Location, ES-Position, ES-CurrentVersion
     98777 BerespHeader   b Location: https://host.es.com:2114/streams/test-1/634
     98777 BerespHeader   b Content-Type: text/plain; charset=utf-8
     98777 BerespHeader   b Server: Mono-HTTPAPI/1.0
     98777 BerespHeader   b Date: Fri, 09 Mar 2018 18:51:11 GMT
     98777 BerespHeader   b Transfer-Encoding: chunked
     98777 BerespHeader   b Keep-Alive: timeout=15,max=100
     98777 VCL_call       b BACKEND_RESPONSE
     98777 TTL            b VCL -1 120 0 1520621471
     98777 BerespHeader   b X-Cacheable: YES
     98777 VCL_return     b deliver
     98777 Storage        b malloc Transient
     98777 ObjProtocol    b HTTP/1.1
     98777 ObjStatus      b 201
     98777 ObjReason      b Created
     98777 ObjHeader      b Access-Control-Allow-Methods: POST, DELETE, GET, OPTIONS
     98777 ObjHeader      b Access-Control-Allow-Headers: Content-Type, X-Requested-With, X-Forwarded-Host, X-Forwarded-Prefix, X-PINGOTHER, Authorization, ES-LongPoll, ES-ExpectedVersion, ES-EventId, ES-EventType, ES-RequiresMaster, ES-HardDelete, ES-ResolveLinkTos
     98777 ObjHeader      b Access-Control-Allow-Origin: *
     98777 ObjHeader      b Access-Control-Expose-Headers: Location, ES-Position, ES-CurrentVersion
     98777 ObjHeader      b Location: https://host.es.com:2114/streams/test-1/634
     98777 ObjHeader      b Content-Type: text/plain; charset=utf-8
     98777 ObjHeader      b Server: Mono-HTTPAPI/1.0
     98777 ObjHeader      b Date: Fri, 09 Mar 2018 18:51:11 GMT
     98777 ObjHeader      b X-Cacheable: YES
     98777 Fetch_Body     b 2 chunked stream
     98777 BackendReuse   b 29 boot.es_dir({ES_MASTER})
     98777 Timestamp      b BerespBody: 1520621471.123029 0.037216 0.000092
     98777 Length         b 0
     98777 BereqAcct      b 463 161 624 679 0 679
     98777 End            b
     98776 Begin          c req 98775 rxreq
     98776 Timestamp      c Start: 1520621471.085746 0.000000 0.000000
     98776 Timestamp      c Req: 1520621471.085746 0.000000 0.000000
     98776 ReqStart       c  {lb} 61133
     98776 ReqMethod      c POST
     98776 ReqURL         c /streams/test-1
     98776 ReqProtocol    c HTTP/1.1
     98776 ReqHeader      c Host: host.es.com
     98776 ReqHeader      c Authorization: Basic ====
     98776 ReqHeader      c Content-Type: application/vnd.eventstore.events+json; charset=utf-8
     98776 ReqHeader      c Accept: application/vnd.eventstore.atom+json; charset=utf-8
     98776 ReqHeader      c Transfer-Encoding: chunked
     98776 ReqHeader      c X-Cloud-Trace-Context: ZZZZZZZ1/12342541254
     98776 ReqHeader      c Via: 1.1 google
     98776 ReqHeader      c X-Forwarded-For: 8.X.X.X, 8.X.X.X
     98776 ReqHeader      c X-Forwarded-Proto: https
     98776 ReqHeader      c Connection: Keep-Alive
     98776 ReqUnset       c X-Forwarded-For: 8.X.X.X, 8.X.X.X
     98776 ReqHeader      c X-Forwarded-For: 8.X.X.X, 8.X.X.X, {lb}
     98776 VCL_call       c RECV
     98776 ReqHeader      c ES-TrustedAuth: anonymous
     98776 ReqUnset       c Authorization: Basic ====
     98776 VCL_return     c pass
     98776 VCL_call       c HASH
     98776 VCL_return     c lookup
     98776 VCL_call       c PASS
     98776 VCL_return     c fetch
     98776 Link           c bereq 98777 pass
     98776 Storage        c malloc Transient
     98776 Timestamp      c ReqBody: 1520621471.086982 0.001236 0.001236
     98776 Timestamp      c Fetch: 1520621471.123026 0.037280 0.036044
     98776 RespProtocol   c HTTP/1.1
     98776 RespStatus     c 201
     98776 RespReason     c Created
     98776 RespHeader     c Access-Control-Allow-Methods: POST, DELETE, GET, OPTIONS
     98776 RespHeader     c Access-Control-Allow-Headers: Content-Type, X-Requested-With, X-Forwarded-Host, X-Forwarded-Prefix, X-PINGOTHER, Authorization, ES-LongPoll, ES-ExpectedVersion, ES-EventId, ES-EventType, ES-RequiresMaster, ES-HardDelete, ES-ResolveLinkTos
     98776 RespHeader     c Access-Control-Allow-Origin: *
     98776 RespHeader     c Access-Control-Expose-Headers: Location, ES-Position, ES-CurrentVersion
     98776 RespHeader     c Location: https://host.es.com:2114/streams/test-1/634
     98776 RespHeader     c Content-Type: text/plain; charset=utf-8
     98776 RespHeader     c Server: Mono-HTTPAPI/1.0
     98776 RespHeader     c Date: Fri, 09 Mar 2018 18:51:11 GMT
     98776 RespHeader     c X-Cacheable: YES
     98776 RespHeader     c X-Varnish: 98776
     98776 RespHeader     c Age: 0
     98776 RespHeader     c Via: 1.1 varnish (Varnish/5.2)
     98776 VCL_call       c DELIVER
     98776 RespHeader     c X-Varnish-Cache: MISS
     98776 RespHeader     c X-Varnish-IP: {varnish}
     98776 VCL_return     c deliver
     98776 Timestamp      c Process: 1520621471.123057 0.037311 0.000031
     98776 RespHeader     c Content-Length: 0
     98776 RespHeader     c Connection: keep-alive
     98776 Timestamp      c Resp: 1520621471.123110 0.037364 0.000053
     98776 ReqAcct        c 473 161 634 787 0 787
     98776 End            c
```

varnish request for a request passed to a slave and forwarded to master:
```
    131335 Begin          c sess 0 HTTP/1
    131335 SessOpen       c {lb} 63804 a0 {varnish} 80 1520621471.613578 26
    131335 Link           c req 131336 rxreq
    131337 Begin          b bereq 131336 pass
    131337 Timestamp      b Start: 1520621471.613964 0.000000 0.000000
    131337 BereqMethod    b POST
    131337 BereqURL       b /streams/test-1
    131337 BereqProtocol  b HTTP/1.1
    131337 BereqHeader    b Host: host.es.com
    131337 BereqHeader    b Content-Type: application/vnd.eventstore.events+json; charset=utf-8
    131337 BereqHeader    b Accept: application/vnd.eventstore.atom+json; charset=utf-8
    131337 BereqHeader    b X-Cloud-Trace-Context:  ZZZZZZZ2/12345566
    131337 BereqHeader    b Via: 1.1 google
    131337 BereqHeader    b X-Forwarded-Proto: https
    131337 BereqHeader    b X-Forwarded-For: 8.X.X.X, 8.X.X.X, {lb}
    131337 BereqHeader    b ES-TrustedAuth: anonymous
    131337 BereqHeader    b X-Varnish: 131337
    131337 VCL_call       b BACKEND_FETCH
    131337 VCL_return     b fetch
    131337 BackendOpen    b 25 boot.es_dir({ES_SLAVE}) {ES_SLAVE} 2114 {varnish} 42010
    131337 BereqHeader    b Transfer-Encoding: chunked
    131337 BackendStart   b 10.20.12.17 2114
    131337 Timestamp      b Bereq: 1520621471.614068 0.000104 0.000104
    131337 Timestamp      b Beresp: 1520621471.617800 0.003836 0.003732
    131337 BerespProtocol b HTTP/1.1
    131337 BerespStatus   b 415
    131337 BerespReason   b Invalid or missing Content-Type
    131337 BerespHeader   b Access-Control-Allow-Methods:
    131337 BerespHeader   b Access-Control-Allow-Headers: Content-Type, X-Requested-With, X-Forwarded-Host, X-Forwarded-Prefix, X-PINGOTHER, Authorization, ES-LongPoll, ES-ExpectedVersion, ES-EventId, ES-EventType, ES-RequiresMaster, ES-HardDelete, ES-ResolveLinkTos
    131337 BerespHeader   b Access-Control-Allow-Origin: *
    131337 BerespHeader   b Access-Control-Expose-Headers: Location, ES-Position, ES-CurrentVersion
    131337 BerespHeader   b Server: Mono-HTTPAPI/1.0
    131337 BerespHeader   b Date: Fri, 09 Mar 2018 18:51:11 GMT
    131337 BerespHeader   b Content-Length: 0
    131337 BerespHeader   b Keep-Alive: timeout=15,max=99
    131337 VCL_call       b BACKEND_RESPONSE
    131337 TTL            b VCL -1 120 0 1520621472
    131337 BerespHeader   b X-Cacheable: YES
    131337 VCL_return     b deliver
    131337 Storage        b malloc Transient
    131337 ObjProtocol    b HTTP/1.1
    131337 ObjStatus      b 415
    131337 ObjReason      b Invalid or missing Content-Type
    131337 ObjHeader      b Access-Control-Allow-Methods:
    131337 ObjHeader      b Access-Control-Allow-Headers: Content-Type, X-Requested-With, X-Forwarded-Host, X-Forwarded-Prefix, X-PINGOTHER, Authorization, ES-LongPoll, ES-ExpectedVersion, ES-EventId, ES-EventType, ES-RequiresMaster, ES-HardDelete, ES-ResolveLinkTos
    131337 ObjHeader      b Access-Control-Allow-Origin: *
    131337 ObjHeader      b Access-Control-Expose-Headers: Location, ES-Position, ES-CurrentVersion
    131337 ObjHeader      b Server: Mono-HTTPAPI/1.0
    131337 ObjHeader      b Date: Fri, 09 Mar 2018 18:51:11 GMT
    131337 ObjHeader      b Content-Length: 0
    131337 ObjHeader      b X-Cacheable: YES
    131337 Fetch_Body     b 0 none -
    131337 BackendReuse   b 25 boot.es_dir({ES_SLAVE})
    131337 Timestamp      b BerespBody: 1520621471.617853 0.003889 0.000053
    131337 Length         b 0
    131337 BereqAcct      b 464 161 625 538 0 538
    131337 End            b
    131336 Begin          c req 131335 rxreq
    131336 Timestamp      c Start: 1520621471.613881 0.000000 0.000000
    131336 Timestamp      c Req: 1520621471.613881 0.000000 0.000000
    131336 ReqStart       c {lb} 63804
    131336 ReqMethod      c POST
    131336 ReqURL         c /streams/test-1
    131336 ReqProtocol    c HTTP/1.1
    131336 ReqHeader      c Host: host.es.com
    131336 ReqHeader      c Authorization: Basic ====
    131336 ReqHeader      c Content-Type: application/vnd.eventstore.events+json; charset=utf-8
    131336 ReqHeader      c Accept: application/vnd.eventstore.atom+json; charset=utf-8
    131336 ReqHeader      c Transfer-Encoding: chunked
    131336 ReqHeader      c X-Cloud-Trace-Context: ZZZZZZZ2/12345566
    131336 ReqHeader      c Via: 1.1 google
    131336 ReqHeader      c X-Forwarded-For: 8.X.X.X, 8.X.X.X
    131336 ReqHeader      c X-Forwarded-Proto: https
    131336 ReqHeader      c Connection: Keep-Alive
    131336 ReqUnset       c X-Forwarded-For: 8.X.X.X, 8.X.X.X
    131336 ReqHeader      c X-Forwarded-For: 8.X.X.X, 8.X.X.X, {lb}
    131336 VCL_call       c RECV
    131336 ReqHeader      c ES-TrustedAuth: anonymous
    131336 ReqUnset       c Authorization: Basic ====
    131336 VCL_return     c pass
    131336 VCL_call       c HASH
    131336 VCL_return     c lookup
    131336 VCL_call       c PASS
    131336 VCL_return     c fetch
    131336 Link           c bereq 131337 pass
    131336 Storage        c malloc Transient
    131336 Timestamp      c ReqBody: 1520621471.614053 0.000172 0.000172
    131336 Timestamp      c Fetch: 1520621471.617876 0.003995 0.003823
    131336 RespProtocol   c HTTP/1.1
    131336 RespStatus     c 415
    131336 RespReason     c Invalid or missing Content-Type
    131336 RespHeader     c Access-Control-Allow-Methods:
    131336 RespHeader     c Access-Control-Allow-Headers: Content-Type, X-Requested-With, X-Forwarded-Host, X-Forwarded-Prefix, X-PINGOTHER, Authorization, ES-LongPoll, ES-ExpectedVersion, ES-EventId, ES-EventType, ES-RequiresMaster, ES-HardDelete, ES-ResolveLinkTos
    131336 RespHeader     c Access-Control-Allow-Origin: *
    131336 RespHeader     c Access-Control-Expose-Headers: Location, ES-Position, ES-CurrentVersion
    131336 RespHeader     c Server: Mono-HTTPAPI/1.0
    131336 RespHeader     c Date: Fri, 09 Mar 2018 18:51:11 GMT
    131336 RespHeader     c Content-Length: 0
    131336 RespHeader     c X-Cacheable: YES
    131336 RespHeader     c X-Varnish: 131336
    131336 RespHeader     c Age: 0
    131336 RespHeader     c Via: 1.1 varnish (Varnish/5.2)
    131336 VCL_call       c DELIVER
    131336 RespHeader     c X-Varnish-Cache: MISS
    131336 RespHeader     c X-Varnish-IP: {varnish}
    131336 VCL_return     c deliver
    131336 Timestamp      c Process: 1520621471.617887 0.004006 0.000012
    131336 RespHeader     c Connection: keep-alive
    131336 Timestamp      c Resp: 1520621471.617913 0.004032 0.000025
    131336 ReqAcct        c 473 161 634 656 0 656
    131336 End            c
```

**Comment by fhoutr01** [comment]
We just recently ran into this as well. If it wasn't for your post we'd still be trying to find this "ghost".
Thanks for the ticket & workaround.

I hope they can get it fixed soon..

**Comment by oskardudycz** [comment]
@hayley-jean, is this something that we're aware of and planning to tackle?

**Comment by hayley-jean** [comment]
This should be fixed in versions 20.10 and greater as we don't forward writes between nodes over HTTP any more.
